securedrop/tests/test_2fa.py

```diff
@@ -1,17 +1,73 @@
 # -*- coding: utf-8 -*-
 import os
+import pytest
+import time
 
+from contextlib import contextmanager
+from datetime import datetime, timedelta
 from flask import url_for
+from pyotp import TOTP
 import flask_testing
 
 os.environ['SECUREDROP_ENV'] = 'test'  # noqa
 from models import Journalist, BadTokenException
+from utils.instrument import InstrumentedApp
 
 import models
 import journalist
 import utils
 
 
+@contextmanager
+def totp_window():
+    # To ensure we have enough time during a single TOTP window to do the
+    # whole test, optionally sleep.
+    now = datetime.now()
+    mod = now.second % 30
+    if mod < 3:
+        time.sleep(mod % 30)
+        now = datetime.now()
+
+    yield
+
+    # This check ensures that the token was used during the same window
+    # in the event that the app's logic only checks for token reuse if the
+    # token was valid.
+    then = datetime.now()
+    assert then < now + timedelta(seconds=(30 - mod))
```

Comment:
No, but we want to test that a token isn't reused, and it needs to happen in the 30 second window for the token to be the same. Also the order we check in the code is 

1. Is it valid?
2. Is it new?

We could mock the uniqueness check to ensure that it was called to protect against regressions that aren't caught by a refactor, but I think it's better to use the context manager because it's less brittle in a sense.

Reply:
In my tor browser based testing branch, I am actually waiting and then using the new token generated to verify the real life usecase.

Reply:
I think that means your tests are testing the wrong thing. The window is only to allow some wiggle room because of high latency with Tor. This is to provide a less frustrating UX. Like I said above, testing token `123456` during `window_1` and then again during `window_2` might short circuit because the `is_token_valid` check aborts early and we never hit the code that checks if a token was reused.

For this test to truly check the behavior we want, it *must* happen in during the same 30 second window to ensure that the token is the same and no short circuit happens.